Add valkyrie IK example

pangtao22 opened the pull request:
finds a standing posture of Valkyrie by solving IK and displays it in drake visualizer
This is my first ever PR. Thank you for your comments!

@hongkai-dai for feature review.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/4009)
<!-- Reviewable:end -->

RussTedrake commented:
@pangtao22 -- great to see.  but your git history contains all of the urdfs, meshes, etc that we wanted to avoid.  can you start a new branch (from master), copy over just the two files, and PR just a single commit?

hongkai-dai commented:
Review status: 0 of 2 files reviewed at latest revision, 27 unresolved discussions.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 3 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVejZJ_Zquz08u0AgUG:-KVejZJ_Zquz08u0AgUH:b-ersixy) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L3)):*
> ```cpp
> // finds a standing posture of Valkyrie by solving IK and displays it in drake
> // visualizer
> ```

Could you move this file to Valkyrie/test folder?

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 15 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVenQeTvjSACNavs8tl:-KVenQeTvjSACNavs8tm:b-hr0bxz) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L15)):*
> ```cpp
> #include "drake/systems/plants/rigid_body_plant/drake_visualizer.h"
> 
> // Includes for IK solver
> ```

punctuation at the end of the sentence.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 16 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVejf-tKBCY2AVQJLBa:-KVejf-tKBCY2AVQJLBb:b-lxl97w) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L16)):*
> ```cpp
> 
> // Includes for IK solver
> #include "drake/systems/plants/IKoptions.h"
> ```

I recommend to move IKoptions.h,, RigidBodyIK.h and RigidBodyConstraint.h to the top of the included files. This test is meant for these IK files.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 36 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVek8Hwq-CzVbT6zXDM:-KVek8Hwq-CzVbT6zXDN:bh1qlsh) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L36)):*
> ```cpp
>  * searched).
>  */
> std::vector<int> GetJointPositionVectorIndices(const RigidBodyTree* tree,
> ```

Is file function copied from somewhere else? This looks like a utility function that should live in RigidBodyTree.h, and to be used by other functions also.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 50 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVekmH16OZlSTeBi-mk:-KVekmH2VmAv2ZxNETYx:beethvc) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L50)):*
> ```cpp
> }
> 
> void findJointAndInsert(const RigidBodyTree* model, const std::string& name,
> ```

change function name to `FindJointAndInsert`.  Also add the documentation for this function.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 58 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVel2Tfo1p10s57kBe1:-KVel2Tfo1p10s57kBe2:bbx49xe) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L58)):*
> ```cpp
> }
> 
> int do_main() {
> ```

I recommend to use GTEST_TEST, instead of this do_main.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 64 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVel8Jdy9vkrY7tYy1F:-KVel8Jdy9vkrY7tYy1G:bs2hw76) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L64)):*
> ```cpp
>           "valkyrie_A_sim_drake_one_neck_dof_wide_ankle_rom.urdf",
>       systems::plants::joints::kRollPitchYaw);
>   // tests
> ```

Delete the commented block.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 90 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVelC9VRc5sGKd5I9Yu:-KVelC9VRc5sGKd5I9Yv:b-vulhj2) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L90)):*
> ```cpp
>   */
> 
>   // setting up constraints, based on testIKMoreConstraints.cpp and
> ```

Setting up, capital case.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 152 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVelmRXD7PGV-wxNV6N:-KVelmRXD7PGV-wxNV6O:b8ljhly) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L152)):*
> ```cpp
>   // constraints are imposed on frames/bodies
>   int l_foot = tree->FindBodyIndex("leftFoot");
>   // Vector4d lfoot_quat = tree->relativeQuaternion(cache, l_foot, 0);
> ```

Delete this commented line.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 154 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVelrFFxPldP__7pX55:-KVelrFFxPldP__7pX56:b2nreh6) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L154)):*
> ```cpp
>   // Vector4d lfoot_quat = tree->relativeQuaternion(cache, l_foot, 0);
>   Vector4d lfoot_quat(1, 0, 0, 0);
>   // std::cout << "Relative quaternion between left foot and world" << std::endl
> ```

Same as above.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 158 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVeltjvZbxusFHdNo-C:-KVeltjvZbxusFHdNo-D:b2nreh6) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L158)):*
> ```cpp
> 
>   const Vector3d origin(0, 0, 0);
>   // std::cout << "origin is" << std::endl << origin << std::endl;
> ```

Same as above.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 160 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVelzJb6pLq39PJEZR7:-KVelzJb6pLq39PJEZR8:b2nreh6) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L160)):*
> ```cpp
>   // std::cout << "origin is" << std::endl << origin << std::endl;
>   auto lfoot_pos0 = tree->transformPoints(cache, origin, l_foot, 0);
>   // std::cout << "Relative position between left foot and world" << std::endl
> ```

Same as above.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 162 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVem0yYDGOiuQRhHnir:-KVem0yYDGOiuQRhHnis:b-81ok8k) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L162)):*
> ```cpp
>   // std::cout << "Relative position between left foot and world" << std::endl
>   // << lfoot_pos0 << std::endl;
>   Vector3d lfoot_pos_lb = lfoot_pos0;
> ```

Why it is not an equality constraint, but instead you want to relax the position constraint?

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 172 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVemGoy7EA9QJ3x96ta:-KVemGoy7EA9QJ3x96tb:bfkdsi5) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L172)):*
> ```cpp
>   WorldPositionConstraint kc_lfoot_pos(tree.get(), l_foot, origin, lfoot_pos_lb,
>                                        lfoot_pos_ub, tspan);
>   double tol = 0.0017453292519943296;
> ```

Magic number.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 186 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVemO8So6BFqTfdco-H:-KVemO8So6BFqTfdco-I:b-w2vap4) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L186)):*
> ```cpp
>   // << rfoot_pos0 << std::endl;
>   Vector3d rfoot_pos_lb = rfoot_pos0;
>   rfoot_pos_lb(0) -= 0.0001;
> ```

Same as the left foot position constraint, why do you want to relax it?

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 206 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVemVo-vGfOQtHp66Yx:-KVemVo-vGfOQtHp66Yy:bfkdsi5) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L206)):*
> ```cpp
>   // torso_idx[2] << std::endl;
>   Vector3d torso_nominal = Vector3d::Zero(3);
>   Vector3d torso_half_range(0.2617993877991494, 0.4363323129985824, inf);
> ```

Magic number.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 209 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVemYnEPhdiK_M_yULr:-KVemYnEPhdiK_M_yULs:bfkdsi5) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L209)):*
> ```cpp
>   Vector3d torso_lb = torso_nominal - torso_half_range;
>   Vector3d torso_ub = torso_nominal + torso_half_range;
>   torso_lb(1) = -0.08726646259971647;
> ```

Magic number.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 224 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVemdnwt2_ph1Dni04z:-KVemdnwt2_ph1Dni05-:bfkdsi5) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L224)):*
> ```cpp
>   // knee_lb(0) += 0.60;
>   Vector2d knee_ub = knee_nominal;
>   knee_ub(0) += 1.90;
> ```

Magic number.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 242 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVeo6nkLuO0KQ6JiAhq:-KVeo6nkLuO0KQ6JiAhr:b-gmw580) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L242)):*
> ```cpp
>   // std::cout << "Number of elements in larm_idx " << larm_idx.size() <<
>   // std::endl;
>   VectorXd larm_lb = VectorXd::Zero(7);
> ```

`Eigen::Matrix<double, 7, 1>`, avoid allocating memory on heap.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 266 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVemoZ2mHvZjtdUmSNH:-KVemoZ2mHvZjtdUmSNI:brr19ms) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L266)):*
> ```cpp
>   // 8 quasistatic constraint
>   QuasiStaticConstraint kc_quasi(tree.get(), tspan);
>   kc_quasi.setShrinkFactor(0.2);
> ```

0.2 corresponds to a very small CoM projection region, why you want this small number?

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 271 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVenlaIqPIhZ7LL5G8j:-KVenlaIqPIhZ7LL5G8k:b-d4tjtg) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L271)):*
> ```cpp
>   auto leftFootPtr = tree->FindBody("leftFoot");
>   Matrix3Xd leftFootContactPts = leftFootPtr->get_contact_points();
>   Matrix3Xd l_foot_pts = leftFootContactPts.rightCols(8);
> ```

`Eigen::Matrix<double, 3, 8>` and `rightCols<8>()`, avoid allocating memory on heap.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 276 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVensv6Lav3No2zSwuU:-KVensv6Lav3No2zSwuV:b-du3ch7) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L276)):*
> ```cpp
>   auto rightFootPtr = tree->FindBody("rightFoot");
>   Matrix3Xd rightFootContactPts = rightFootPtr->get_contact_points();
>   Matrix3Xd r_foot_pts = rightFootContactPts.rightCols(8);
> ```

rightCols<8>().

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 279 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVen0UKmnDaBsCrhcyp:-KVen0UKmnDaBsCrhcyq:b-4o1ogh) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L279)):*
> ```cpp
>   kc_quasi.addContact(1, &r_foot, &r_foot_pts);
> 
>   std::cout << "LeftFootContactPts:" << std::endl;
> ```

Remove these cout.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 310 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVen5DyFnrGaCKpMjCX:-KVen5DyFnrGaCKpMjCY:b4orpf4) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L310)):*
> ```cpp
>              constraint_array.data(), ikoptions, &q_sol, &info,
>              &infeasible_constraint);
>   printf("info = %d\n", info);
> ```

Remove printf

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 313 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVen6tgvZunLLQ_yckS:-KVen6tgvZunLLQ_yckT:b-6x7ng4) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L313)):*
> ```cpp
> 
>   /////////////////////////////////////////
>   std::cout
> ```

Remove cout.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 320 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVen8uBqXPgBkxebzSO:-KVen8uBqXPgBkxebzSP:b-m9g963) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L320)):*
> ```cpp
>   printf("%5.6f\n%5.6f\n%5.6f\n", com(0), com(1), com(2));
> 
>   // cout for debugging
> ```

Delete this commented block.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 354 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVenKEFahEeSNEAU4_I:-KVenKEFahEeSNEAU4_J:bcbulxk) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L354)):*
> ```cpp
>   //  diagram->EvalOutput(*context, output.get());
>   diagram->Publish(*context);
>   // systems::Simulator<double> simulator(*diagram);
> ```

Remove commented lines.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4009)*
<!-- Sent from Reviewable.io -->

pangtao22 commented:
Sure. I'll open a new PR after addressing issues raised by Hongkai, so that I don't lose his comments. 

---

Review status: 0 of 2 files reviewed at latest revision, 27 unresolved discussions.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 3 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVejZJ_Zquz08u0AgUG:-KVflAMbXLXXvV5Vuymb:b-7j86r0) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L3)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> Could you move this file to Valkyrie/test folder?
</details>
sure.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 36 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVek8Hwq-CzVbT6zXDM:-KVfTbjpltLSBlW9kbfx:b-kguae8) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L36)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> Is file function copied from somewhere else? This looks like a utility function that should live in RigidBodyTree.h, and to be used by other functions also.
</details>
Yes. These functions are from this test file (with some modifications): https://github.com/RobotLocomotion/drake/blob/c81ae3ea2d3c920188c73d0cffd0cc058636e2fc/drake/systems/plants/test/testIKMoreConstraints.cpp 

Is there a better way to "borrow" these functions other than copying them? 

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 64 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVel8Jdy9vkrY7tYy1F:-KVfUwY0bl2dQBE2uPRE:b-sm3ies) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L64)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> Delete the commented block.
</details>
But the commented block will be useful for debugging when I try to different postures in the future. Should I only keep them in my working branch?

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 152 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVelmRXD7PGV-wxNV6N:-KVfVjX6xjvbV5NtveEF:bdlna6m) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L152)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> Delete this commented line.
</details>
This commented line is an alternative way to construct the constraint, which might be useful to find other postures. Keep them in my working branch only?

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 162 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVem0yYDGOiuQRhHnir:-KVfhXegyhurfMLHRSKR:b-dttmwm) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L162)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> Why it is not an equality constraint, but instead you want to relax the position constraint?
</details>
These are the constraints used by Lucas/Pat in their director IK interface. Is relaxing an equality constraint a bad thing? 

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 172 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVemGoy7EA9QJ3x96ta:-KVfVHCs4P3idJMAnh0B:b-qgkv1z) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L172)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> Magic number.
</details>
This particular tolerance for quaternion constraint not only appears in director, but also in some drake examples. Should I explain in the comments where it comes from?

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 206 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVemVo-vGfOQtHp66Yx:-KVfW2ZmHW8_uVd5EH_V:b-sxjtff) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L206)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> Magic number.
</details>
These numbers are taken from the MATLAB code generated by Director, the one where the robot's hand can be dragged around and the robot adjusts its posture accordingly. Should I explain how I find those numbers in the comment?

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 209 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVemYnEPhdiK_M_yULr:-KVfkLSpB4YSXcA3AyIa:b-ya81ng) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L209)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> Magic number.
</details>
Same as above

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 224 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVemdnwt2_ph1Dni04z:-KVfkN80f7UrW7Jqq-Cg:b-ya81ng) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L224)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> Magic number.
</details>
Same as above

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 242 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVeo6nkLuO0KQ6JiAhq:-KVfjfTRMkz9ImNEB8ko:b-7adn45) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L242)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> `Eigen::Matrix<double, 7, 1>`, avoid allocating memory on heap.
</details>
Is it because the heap is slower to access? Is it a common practice to put small arrays on the stack?

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 266 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVemoZ2mHvZjtdUmSNH:-KVfirVrQBSINbSpWIj2:bia4irp) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L266)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> 0.2 corresponds to a very small CoM projection region, why you want this small number?
</details>
Again this is the number used in director. 

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 313 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVen6tgvZunLLQ_yckS:-KVfjAbQtMWrYKZpS8ID:b-hm14ms) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L313)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> Remove cout.
</details>
Is cout/printf generally discouraged in test files? Please advise on best practice. Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-:-KVfU9yMpVk_paUKbnl6:b432avo)*
<!-- Sent from Reviewable.io -->

pangtao22 commented:
Review status: 0 of 2 files reviewed at latest revision, 27 unresolved discussions.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 266 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVemoZ2mHvZjtdUmSNH:-KVfo0G9fE7PBB6WHfW_:bsjqdq3) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L266)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary>
> Again this is the number used in director. 
</details>
Could you advise on what is the typical value of the shrink factor?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4009)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 0 of 2 files reviewed at latest revision, 27 unresolved discussions.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 64 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVel8Jdy9vkrY7tYy1F:-KVfmnlaMVzCyLKoAlYk:bgz0f60) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L64)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary>
> But the commented block will be useful for debugging when I try to different postures in the future. Should I only keep them in my working branch?
</details>
You may keep them in your working branch, but I think in the master branch we should not leave a commented block. 

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 162 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVem0yYDGOiuQRhHnir:-KVfn006ESR6StXhOen9:b-6fsz9p) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L162)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary>
> These are the constraints used by Lucas/Pat in their director IK interface. Is relaxing an equality constraint a bad thing? 
</details>
I do not see why you want to relax the constraints. If you do not relax the constraint, does the solver converge?

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 172 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVemGoy7EA9QJ3x96ta:-KVfnAQuOuFGRr79ctzZ:b347c3r) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L172)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary>
> This particular tolerance for quaternion constraint not only appears in director, but also in some drake examples. Should I explain in the comments where it comes from?
</details>
Yes, please explain them. And have you tried to set the tolerance to 0? Does the solver converge?

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 206 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVemVo-vGfOQtHp66Yx:-KVfnMVD6qLGmGuRr0LE:b-fafwpn) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L206)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary>
> These numbers are taken from the MATLAB code generated by Director, the one where the robot's hand can be dragged around and the robot adjusts its posture accordingly. Should I explain how I find those numbers in the comment?
</details>
This is not a meaningful number to the reader. I guess they correspond to 15 degrees and 25 degrees. Please use `15.0/180 * M_PI` and `25.0/180*M_PI` for readability.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 209 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVemYnEPhdiK_M_yULr:-KVfnoxW4AM1ale0JOkB:b-4cl092) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L209)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary>
> Same as above
</details>
Same as above, `5.0/180*M_PI`.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 242 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVeo6nkLuO0KQ6JiAhq:-KVfo6HXQZ4IkkNGvVN_:b58sckh) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L242)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary>
> Is it because the heap is slower to access? Is it a common practice to put small arrays on the stack?
</details>
It is because allocating memory on heap happens at run time, while allocating memory to stack happens during compile time. So by avoiding heap allocation, the program runs faster.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 313 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVen6tgvZunLLQ_yckS:-KVfoK4PX7FEdLA8soez:bp2l34k) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L313)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary>
> Is cout/printf generally discouraged in test files? Please advise on best practice. Thanks!
</details>
Yes it is discouraged in the test file. The CI machine does not check these print message.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4009)*
<!-- Sent from Reviewable.io -->

pangtao22 commented:
Review status: 0 of 2 files reviewed at latest revision, 28 unresolved discussions.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 162 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVem0yYDGOiuQRhHnir:-KVfyrUlF4Ogmgttd9kq:b5r620x) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L162)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> I do not see why you want to relax the constraints. If you do not relax the constraint, does the solver converge?
</details>
Yes the solver converges even if the constraints are not relaxed. In general, are equality constraints preferred over relaxed constraint?

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 172 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVemGoy7EA9QJ3x96ta:-KVfz66MNQVeZ1WF5Yvl:br4pm8k) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L172)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> Yes, please explain them. And have you tried to set the tolerance to 0? Does the solver converge?
</details>
Same as above, is it preferred to use as small a tolerlance value as possible?

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 271 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVenlaIqPIhZ7LL5G8j:-KVfw9CW5YVkuZb1S2e3:b11tzv0) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L271)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> `Eigen::Matrix<double, 3, 8>` and `rightCols<8>()`, avoid allocating memory on heap.
</details>
get_contact_points (in the line above) returns a dynamic-sized matrix.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 277 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVfx9ojZktaGvN1fBw_:-KVfx9ojZktaGvN1fBwa:bcesvec) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L277)):*
> ```cpp
>   Matrix3Xd rightFootContactPts = rightFootPtr->get_contact_points();
>   Matrix3Xd r_foot_pts = rightFootContactPts.rightCols(8);
>   kc_quasi.addContact(1, &r_foot, &r_foot_pts);
> ```

kc_quasi.addContact(.........) expects Matrix3Xd for its last argument. Therefore r_foot_pts has to be dynamic-sized?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4009)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 0 of 3 files reviewed at latest revision, 28 unresolved discussions.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 162 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVem0yYDGOiuQRhHnir:-KVgCFLhtm6KGuRvP5Lg:bk3ag0j) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L162)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary>
> Yes the solver converges even if the constraints are not relaxed. In general, are equality constraints preferred over relaxed constraint?
</details>
I think we want to impose meaningful constraints, I do not understand the motivation to relax the constraint here.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 172 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVemGoy7EA9QJ3x96ta:-KVgCZTLXObBQPyFpp_z:bemrgoq) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L172)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary>
> Same as above, is it preferred to use as small a tolerlance value as possible?
</details>
I think the question is why you want to relax it? Using tol = 0 has a clear meaning, that you do not want to move the orientation of the foot, and the foot will be flat on the ground.

---

*[drake/examples/Valkyrie/valkyrie_IK_test.cc, line 277 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4009#-KVfx9ojZktaGvN1fBw_:-KVgDLMkgF8opCS4i45b:bhqcqx1) ([raw file](https://github.com/robotlocomotion/drake/blob/36e8059bf7039aef024a6b8f404b7112f2817ee5/drake/examples/Valkyrie/valkyrie_IK_test.cc#L277)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary>
> kc_quasi.addContact(.........) expects Matrix3Xd for its last argument. Therefore r_foot_pts has to be dynamic-sized?
</details>
Good point, you are right.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4009)*
<!-- Sent from Reviewable.io -->
